### PR TITLE
Add a command-line interface to the package

### PR DESCRIPTION
**Pull request by acmiyaguchi**
This document describes a scriptable interface for prio. This is the interface that I'm aiming to implement. This should be a sufficient building block for building more complex workflows.

### CODE REVIEW

**Comment by acmiyaguchi** [comment]
I current have the following stubs in place and will flesh them out appropriately. I also have done quite a few small changes which include formatting, so I apologize for the line noise ahead of time.

@wlach would you rather I file a PR for each change, or are you okay reviewing a massive PR that might include many surface level changes? 


```
Usage: prio [OPTIONS] COMMAND [ARGS]...

  Command line utility for prio.

Options:
  --help  Show this message and exit.

Commands:
  aggregate  Generate an aggregate share from a batch of verified SNIPs
  publish    Generate a final aggregate and remap data to a content...
  verify1    Decode a batch of shares
  verify2    Verify a batch of SNIPs

Usage: prio verify1 [OPTIONS]

  Decode a batch of shares

Options:
  --output TEXT          [required]
  --private-key TEXT     [required]
  --public-key-B TEXT    [required]
  --public-key-A TEXT    [required]
  --server-id TEXT       [required]
  --prefix TEXT          [required]
  --bucket TEXT          [required]
  --input-internal TEXT  [required]
  --help                 Show this message and exit.

Usage: prio verify2 [OPTIONS]

  Verify a batch of SNIPs

Options:
  --output TEXT          [required]
  --private-key TEXT     [required]
  --public-key-B TEXT    [required]
  --public-key-A TEXT    [required]
  --server-id TEXT       [required]
  --prefix TEXT          [required]
  --bucket TEXT          [required]
  --input-internal TEXT  [required]
  --input-external TEXT  [required]
  --help                 Show this message and exit.

Usage: prio aggregate [OPTIONS]

  Generate an aggregate share from a batch of verified SNIPs

Options:
  --output TEXT          [required]
  --private-key TEXT     [required]
  --public-key-B TEXT    [required]
  --public-key-A TEXT    [required]
  --server-id TEXT       [required]
  --prefix TEXT          [required]
  --bucket TEXT          [required]
  --input-internal TEXT  [required]
  --input-external TEXT  [required]
  --help                 Show this message and exit.

Usage: prio publish [OPTIONS]

  Generate a final aggregate and remap data to a content blocklist

Options:
  --output TEXT          [required]
  --private-key TEXT     [required]
  --public-key-B TEXT    [required]
  --public-key-A TEXT    [required]
  --server-id TEXT       [required]
  --prefix TEXT          [required]
  --bucket TEXT          [required]
  --input-internal TEXT  [required]
  --input-external TEXT  [required]
  --help                 Show this message and exit.
```

**Comment by acmiyaguchi** [comment]
My `test-integration` script reliably segfaults -- it's time to delve back into the debugger to see what's going wrong.

**Comment by acmiyaguchi** [comment]
@wlach f?

This implements the command-line interface. I have a working script under `scripts/test-integration`. There are a couple of things that I need to implement:

* unit tests for each of the sub-commands in pytest
* schemas describing the input and output for each command

During development, I ran into a couple of things that led me astray. The first are segfaults when the wrong data is provided to the functions. The second is that the verifier object must be loaded with the shares (stored under `raw/`) during the verify1, verify2, and aggregate steps or they will fail validation during `aggregate`. Finally, while most steps are order independent (data for server a or server b may be passed in, but in either order), the order in the publish step is significant.

The logs can be found here: https://gist.github.com/acmiyaguchi/da9f18039d0d26ccbd935094ab1296af

The final tree looks like the following:

```
$ tree working/
working/
├── client
│   └── data.ndjson
├── server_a
│   ├── intermediate
│   │   ├── external
│   │   │   ├── aggregate
│   │   │   │   └── data.ndjson
│   │   │   ├── verify1
│   │   │   │   └── data.ndjson
│   │   │   └── verify2
│   │   │       └── data.ndjson
│   │   └── internal
│   │       ├── aggregate
│   │       │   └── data.ndjson
│   │       ├── verify1
│   │       │   └── data.ndjson
│   │       └── verify2
│   │           └── data.ndjson
│   ├── processed
│   │   └── data.ndjson
│   └── raw
│       └── data.ndjson
└── server_b
    ├── intermediate
    │   ├── external
    │   │   ├── aggregate
    │   │   │   └── data.ndjson
    │   │   ├── verify1
    │   │   │   └── data.ndjson
    │   │   └── verify2
    │   │       └── data.ndjson
    │   └── internal
    │       ├── aggregate
    │       │   └── data.ndjson
    │       ├── verify1
    │       │   └── data.ndjson
    │       └── verify2
    │           └── data.ndjson
    ├── processed
    │   └── data.ndjson
    └── raw
        └── data.ndjson
```

--- 

There are a couple of things that would be good to cover in this review:

* Are the command options sane? In particular, the inputs are assumed to be files, while the outputs are directories. The output files assume the basename of the inputs. However, this leads to all of the downstream files containing the same suffix.
* Is it a safe assumption to assume that batches of data will be processed once? I will construct some edge cases where some clients appear in server a but are missing in server b.

There are a lot of changes in this PR, so feel free to comment on anything or everything.

File: prio/cli/commands.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+    output_a,
+    output_b,
+):
+    public_key_internal = libprio.PublicKey_import_hex(public_key_internal)
+    public_key_external = libprio.PublicKey_import_hex(public_key_external)
+    config = libprio.PrioConfig_new(
+        n_data, public_key_internal, public_key_external, batch_id
+    )
+
+    with open(input, "r") as f:
```
Comment (wlach) [inline prio/cli/commands.py]: nit, can skip the `, "r"`

File: prio/cli/commands.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    shared_secret = b64decode(shared_secret)
+
+    config = libprio.PrioConfig_new(
+        n_data, public_key_internal, public_key_external, batch_id
+    )
+    server = libprio.PrioServer_new(config, server_id, private_key, shared_secret)
+    verifier = libprio.PrioVerifier_new(server)
+    packet = libprio.PrioPacketVerify1_new()
+
+    with open(input, "r") as f:
```
Comment (wlach) [inline prio/cli/commands.py]: same

File: prio/cli/commands.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+    packet1_internal = libprio.PrioPacketVerify1_new()
+    packet1_external = libprio.PrioPacketVerify1_new()
+    packet = libprio.PrioPacketVerify2_new()
+
+    with open(input, "r") as f:
+        data = map(json.loads, f.readlines())
+    with open(input_internal, "r") as f:
+        data_internal = map(json.loads, f.readlines())
+    with open(input_external, "r") as f:
```
Comment (wlach) [inline prio/cli/commands.py]: Same as above, the "r" isn't necessary.

File: prio/cli/commands.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    output,
+):
+    """Decode a batch of shares"""
+    click.echo("Running verify1")
+
+    private_key = libprio.PrivateKey_import_hex(private_key, public_key_internal)
+    public_key_internal = libprio.PublicKey_import_hex(public_key_internal)
+    public_key_external = libprio.PublicKey_import_hex(public_key_external)
+    server_id = libprio.PRIO_SERVER_A if server_id == "A" else libprio.PRIO_SERVER_B
+    shared_secret = b64decode(shared_secret)
```
Comment (wlach) [inline prio/cli/commands.py]: It seems like this block tends to be duplicated across many of the commands. Could you write some small helper functions to do this work, that could be shared? Also probably better not to modify input variables, but rather to transform them.

Could you do something like this:

```py
def import_keys(private_key_hex, public_key_internal_hex, public_key_external_hex):
    return (
        libprio.PrivateKey_import_hex(private_key_hex, public_key_internal_hex),
        libprio.PublicKey_import_hex(public_key_internal_hex),
        libprio.PublicKey_import_hex(public_key_external_hex)
    )
...
(private_key, public_key_internal, public_key_external) = import_keys(private_key_hex, public_key_internal_hex, public_key_external_hex)
```
```

File: scripts/create-folder [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    "intermediate/external/aggregate",
+    "processed/",
+]
+
+for server in ["server_a", "server_b"]:
+    for path in paths:
+        p = os.path.join(workdir, server, path)
+        os.makedirs(p, exist_ok=True)
+
+print(f"Done creating folders in {workdir}")
```
Comment (wlach) [inline scripts/create-folder]: Nit: I would probably skip this output.

File: scripts/test-cli-integration [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+python -m prio encode-shares \
+    --n-data 3 \
+    --batch-id test \
+    --public-key-internal $(jq -r ".public_key" <<< $key_a) \
+    --public-key-external $(jq -r ".public_key" <<< $key_b) \
+    --input working/client/data.ndjson \
+    --output-A working/server_a/raw/ \
+    --output-B working/server_b/raw/
+
+jq -c '.' working/server_a/raw/data.ndjson
```
Comment (wlach) [inline scripts/test-cli-integration]: Could we make things like `working/server_a/raw/` bash variables like:

```bash
SERVER_A_WORKDIR=working/server_a/raw/
```

This makes it a little easier to catch typos and errors.

Comment (acmiyaguchi) [reply scripts/test-cli-integration]: I refactored this to extract the environment variables out, it looks a lot nicer now.

File: prio/cli/commands.py (hunk 4 above)
Comment (acmiyaguchi) [reply prio/cli/commands.py]: Done, I've avoided modifying input variables and did some refactoring.

**Pull request merged**